### ISSUE TITLE
Input type (torch.FloatTensor) and weight type (torch.cuda.FloatTensor) should be the same or input should be a MKLDNN tensor and weight is a dense tensor

### ISSUE THREAD

**Issue by er8235ask** [ultralytics/ultralytics#18877]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Not too sure what happened, I downgraded ultralytics to see something and then my export started getting that error. I then uninstalled and reinstalled the newest version and still got this error. To make sure it wasn't my python installation that got messed up I did it again on a fresh venv which still got the same issue.

```
from ultralytics import YOLO
import yaml

model.export(
    format="engine",
    dynamic=False,  
    batch=1,  
    workspace=20, 
    imgsz=320, 
    int8=True,
    data=data, 
    device=0 
)

```

Which returns:
```
ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: export failure ❌ 0.2s: Input type (torch.FloatTensor) and weight type (torch.cuda.FloatTensor) should be the same or input should be a MKLDNN tensor and weight is a dense tensor
TensorRT: export failure ❌ 0.2s: Input type (torch.FloatTensor) and weight type (torch.cuda.FloatTensor) should be the same or input should be a MKLDNN tensor and weight is a dense tensor
Backend qtagg is interactive backend. Turning interactive mode on.
```

### Environment

```
Ultralytics 8.3.67 🚀 Python-3.12.1 torch-2.5.1+cu118 CUDA:0 (NVIDIA GeForce RTX 3090, 24576MiB)
Setup complete ✅ (12 CPUs, 31.9 GB RAM, 680.6/3717.1 GB disk)

OS                  Windows-11-10.0.22631-SP0
Environment         Windows
Python              3.12.1
Install             pip
RAM                 31.91 GB
Disk                680.6/3717.1 GB
CPU                 AMD Ryzen 5 5600X 6-Core Processor
CPU count           12
GPU                 NVIDIA GeForce RTX 3090, 24576MiB
GPU count           1
CUDA                11.8
```

### Minimal Reproducible Example

```
from ultralytics import YOLO
import yaml

model = YOLO(yolo11n.pt)

model.export(
    format="engine",
    dynamic=True,  
    batch=1,  
    workspace=20, 
    imgsz=320, 
    int8=True,
    data="data.yaml", 
    device=0 
)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @er8235ask, thank you for your interest in Ultralytics 🚀! We appreciate you providing the details of your setup and code. Your issue seems to be related to a 🐛 bug during the export process.

To assist us in debugging this further, please ensure that the provided code is a working [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/). For example, ensure that all variables like `yolo11n.pt` and `"data.yaml"` are defined or replaceable with dummy placeholders. This will allow us to reproduce the issue on our end. 

Meanwhile, please try the following steps in your setup:
1. Upgrade to the latest version of the `ultralytics` package and ensure all dependencies are installed correctly:
   ```bash
   pip install -U ultralytics
   ```
2. Double-check your installed PyTorch and CUDA versions. Based on your `torch-2.5.1+cu118` setup, ensure compatibility with your export task. A clean environment, as you've already tried, is a great step.

Visit the [Docs](https://docs.ultralytics.com/) for additional guidance on export processes and requirements, especially [Exporting Models to Different Formats](https://docs.ultralytics.com/modes/export/).

💡 For further troubleshooting:
- Real-time conversations take place on our [Discord](https://discord.com/invite/ultralytics) channel 🎧.
- In-depth community discussions are available on [Discourse](https://community.ultralytics.com/).
- Or, explore our [Subreddit](https://reddit.com/r/Ultralytics) for community-driven advice and support.

## Verified Environments

If you're running into persistent issues, consider using an up-to-date verified environment:
- <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>  
- <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
- <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  

Alternatively, see our guides for [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [Amazon AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/), or our [Docker](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) setup to simplify environment compatibility.

## Status Check

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, it indicates that all tests in the [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) pipeline are passing—ensuring functionality across platforms.

This is an automated response to guide you while waiting for an Ultralytics engineer to assist you further. Thank you for your patience and for helping us improve! 😊

**Comment by Y-T-G** [comment]
Does it work if you only use `format="engine"` and no other arguments?

**Comment by Y-T-G** [comment]
Does it work with Ultralytics v8.3.66?

**Comment by Y-T-G** [comment]
Is this the pretrained model?

**Comment by Y-T-G** [comment]
Does the pretrained model export fine?

**Comment by Y-T-G** [comment]
You can try using Python 3.11

**Comment by Y-T-G** [comment]
Try `workspace=None`

**Comment by Y-T-G** [comment]
> Worked with workspace none. Is there a reason why it doesn't work with 20? The peak memory usage it prints out isn't even close to my maximum

`None` automatically determines the `workspace`

**Comment by Y-T-G** [comment]
You would need to ask on NVIDIA forums why it doesn't work with explicit workspace since it's an error from TensorRT and not Ultralytics.

**Comment by BitCode-in** [comment]
There is no error on ultralytics version 8.3.23, the last one also had an error.

**Comment by glenn-jocher** [comment]
Thanks for reporting this. For TensorRT export issues related to workspace size or device-specific errors, we recommend:  
1. Use `workspace=None` to allow TensorRT auto-allocation as shown in the TensorRT export documentation  
2. Ensure proper calibration data when using INT8 via the `data` argument  

The warnings during INT8 export are normal TensorRT messages about uncalibrated layers. For detailed TensorRT behavior, you may want to consult the NVIDIA TensorRT Developer Guide since these are engine-level messages.

**Comment by Y-T-G** [comment]
@BitCode-in You can try this

https://github.com/ultralytics/ultralytics/pull/19541/

**Comment by BitCode-in** [comment]
@Y-T-G 

### Version 8.3.84 (ERROR)❌

```
yolo export model=yolo11n.pt format=engine dynamic=True workspace=2

WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.3.84 🚀 Python-3.8.5 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3050 Laptop GPU, 4096MiB)
.......
TensorRT: input "images" with shape(-1, 3, -1, -1) DataType.FLOAT
TensorRT: output "output0" with shape(-1, 84, -1) DataType.FLOAT
TensorRT: WARNING ⚠️ 'dynamic=True' model requires max batch size, i.e. 'batch=16'
TensorRT: building FP32 engine as yolo11n.engine
[03/06/2025-14:07:29] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[03/06/2025-14:07:29] [TRT] [E] [shapeContext.cpp::nvinfer1::builder::ShapeContext::checkVolume::3856] Error Code 2: Internal Error (Assertion interval.max() >= 0 failed. Condition::kVOLUME_BOUND is undefined for factor in interval [-171798691840,-4])
TensorRT: export failure ❌ 15.0s: __enter__
```


### Version 8.3.23 (succses)✅ 

```
yolo export model=yolo11n.pt format=engine dynamic=True workspace=2

WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.3.23 🚀 Python-3.8.5 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3050 Laptop GPU, 4096MiB)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
........
TensorRT: export success ✅ 105.8s, saved as 'yolo11n.engine' (17.0 MB)

Export complete (109.8s)
```

**Comment by Y-T-G** [comment]
@BitCode-in Is that with the PR?

**Comment by BitCode-in** [comment]
@Y-T-G 
If you don't need the new features in version 8.3.84, try downgrading it to 8.3.23.
pip install ultralytics==8.3.23

**Comment by Y-T-G** [comment]
@BitCode-in You can test the PR with

`pip install git+https://github.com/ultralytics/ultralytics@tensorrt-shape`

**Comment by BitCode-in** [comment]
@Y-T-G Exported
```
yolo export model=yolo11n.pt format=engine dynamic=True workspace=2

WARNING ⚠️ TensorRT requires GPU export, automatically assigning device=0
Ultralytics 8.3.84 🚀 Python-3.8.5 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3050 Laptop GPU, 4096MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
.................
TensorRT: export success ✅ 107.0s, saved as 'yolo11n.engine' (17.5 MB)
```
